DiamondLightSource/hyperion
# Add plan to align DCM and mirrors when changing energy

Fixes #986

Link to dodal PR (if required): https://github.com/DiamondLightSource/dodal/pull/243

### To test:
The code is intended to implement what currently happens in GDA in `align_beam_specific_worker()` between the 
`disableFeedbackLoop()` and `reenableFeedbackLoopIfNeeded()`

* The unit tests have some spurious failures, I've raised #998 and #999

There are a number of implementation issues/questions in the code marked with `###`, namely:
* Initial DCM peak auto adjustment search appears to be centred on around the current pitch position, but I'm not sure where/whether this is actually initialised anywhere?
* There's some code to set the attenuator transmission to 1 and restore it afterwards, I'm not sure if this is something that should be the responsibility of this code, it wasn't mentioned in the parent github issue, so I haven't implemented it, but do you think I should?
* There's a random wait for 2 seconds after the DCM pitch adjusts, but I'm not sure what this is waiting for, is it sufficient to just wait for the group to finish moving?
* HFM initial preset position and search start appear to differ slightly in GDA, but I don't know if this is a deliberate thing or not.
* I couldn't use bps.mv in the unit test for `_adjust_dcm_roll` because the fake EpicsMotor doesn't seem to have a RealPosition attribute, so had to use set (bluesky docs suggest that `mv` is equivalent to `set` with `wait=True`, but looking at the code, this isn't actually the case and `bps.mv` appears to try clever things), is this a known thing?
* The roll angle lookup table code that I wrote, currently does a very primitive peak find which finds the position of the maximum value in the scan. In GDA it does extrapolation from the LUT, the simple numpy linear interpolation implementation doesn't, will this be a problem? Also when I looked at the table in GDA deployed I03 the fit appeared to be completely flat i.e. generated a constant roll angle regardless of Bragg input.
* Originally I thought the gaussian fit wasn't actually called for I03, however as I write this, I now think that maybe it does, so perhaps will have to implement that in another PR. 
* Also I have just noticed in the gaussian code, there is this comment "special" code for I03/04 (depending on function you look at) :
`drive back past peak to start before moving to peak`. This sounds like it's for correcting backlash in mechanism and possibly a good idea, is this something I should do generally?
* Also, would like to try some of this at some point on actual beamline hardware if that's possible to see if it really works!
